Enable pt2e quantization path for arm

choudhary-devang opened the pull request:
**Title**: Enable PyTorch 2 Export Quantization path for ARM CPUs.

**Description:**
 - This PR extends the PyTorch 2 Export Quantization (PT2E Quantization) workflow—originally available only on x86 CPUs—to support ARM platforms. PT2E Quantization is an automated, full-graph quantization solution in PyTorch that improves on Eager Mode Quantization by adding support for functionals and automating the overall process. It is part of the torch.ao module and fully supports quantization when using the compile mode.

**Key Changes:**

 - Introduces ARM-specific support by leveraging oneDNN kernels for matmuls and convolution.

 - Integrates pre-defined configuration selection to automatically choose the best quantization settings based on the selected quantization method.

**Provides customization options via two flags:**

 - **qat_state:** Indicates whether to use Quantization Aware Training (if set to True) or Post Training Quantization (if set to False). The default remains False.
 - **dynamic_state:** Selects between dynamic quantization (if True) and static quantization (if False). The default is also set to False.
![Screenshot 2025-01-22 105543](https://github.com/user-attachments/assets/c611a1ce-9274-4b70-9c58-cae96000d06d)

These options allow users to tailor the quantization process for their specific workload requirements (e.g., using QAT for fine-tuning or PTQ for calibration-based quantization).

Testing and Validation:

The new ARM flow has been thoroughly tested across a range of models with all combinations:
**NLP**: Models such as BERT and T5.
**Vision**: Models like ResNet and ViT.
**Custom Models**: user defined models with various operators.

example script:
```
import torch
import torchvision.models as models
from torch.ao.quantization.quantize_pt2e import prepare_pt2e, convert_pt2e
import torch.ao.quantization.quantizer.arm_inductor_quantizer as armiq
from torch.ao.quantization.quantizer.arm_inductor_quantizer import ArmInductorQuantizer
from torch.profiler import profile, record_function, ProfilerActivity

model_name = "resnet50"
model = models.__dict__[model_name](pretrained=True)

# Set the model to eval mode
model = model.eval()

# Create the data, using the dummy data here as an example
traced_bs = 500
x = torch.randn(traced_bs, 3, 224, 224).contiguous(memory_format=torch.channels_last)
example_inputs = (x,)

with torch.no_grad():
    exported_model = torch.export.export_for_training(model, example_inputs).module()
    quantizer = armiq.ArmInductorQuantizer()
    quantizer.set_global(armiq.get_default_arm_inductor_quantization_config(is_dynamic=False))
    prepared_model = prepare_pt2e(exported_model, quantizer)
    converted_model = convert_pt2e(prepared_model)

    with torch.set_grad_enabled(False):
        for _ in range(50):
            converted_model(*example_inputs) #Warmup
        print("Warmup over")
        with profile(activities=[ProfilerActivity.CPU], record_shapes=True) as prof:
            with record_function("model_inference"):
                for _ in range(100):
                    converted_model(*example_inputs)

    print(prof.key_averages(group_by_input_shape=True).table(sort_by="self_cpu_time_total"))

```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @malfet @snadampal @milpuz01 @aditew01 @nikhil-arm @fadara01

choudhary-devang committed:
Enable pt2e quantization path for arm

maajidkhann commented:
@pytorchbot label "module: arm"

maajidkhann commented:
@pytorchbot label "module: cpu"

maajidkhann commented:
@pytorchbot  label "ciflow/linux-aarch64"

maajidkhann commented:
> @jerryzh168 can you please review this pr, thankyou.

cc @digantdesai @jianyuh @malfet

jerryzh168 reviewed (approved):
thanks, the quantizer can be owned by ARM I think, so LGTM. can you add some tests, similar to https://github.com/pytorch/pytorch/blob/main/test/quantization/pt2e/test_x86inductor_quantizer.py ?

choudhary-devang committed:
added the test_setup

choudhary-devang commented:
Hi @jerryzh168, thanks for the quick response. I added the tests for the arm_inductor_quantizer config.

jerryzh168 commented on c10/core/QEngine.h:
```diff
@@ -17,14 +17,15 @@ enum class QEngine : uint8_t {
   QNNPACK = 2,
   ONEDNN = 3,
   X86 = 4,
+  Arm = 5,
```
are these needed for pt2e quant stack? I feel these are only needed for the older fx stack

Xia-Weiwen replied:
Yes, I think it's only needed for old stacks.

Xia-Weiwen replied:
Same for qconfig.

leslie-fang-intel commented on torch/ao/quantization/qconfig.py:
```diff
@@ -264,6 +265,13 @@ def get_default_qconfig(backend="x86", version=0):
         qconfig
     """
+    # qconfig check for arm
+    if platform.processor() == "arm" and backend in ["fbgemm", "x86", "qnnpack"]:
```
Why this code changes are needed for PT2E quantization?

choudhary-devang commented:
Hi [@huydhn](https://github.com/huydhn), when i am trying to reply to a comment then it showing me the pending label 
![image](https://github.com/user-attachments/assets/ac00e82b-9d99-4b19-b810-19da022d7988)

![image](https://github.com/user-attachments/assets/0b2ebc9f-f8c0-47c2-9327-a9836cfb5717)

so the other reviewers are not able to see my comments, can you help me with this.

choudhary-devang commented:
Hi @jerryzh168 , @Xia-Weiwen, replay for this (https://github.com/pytorch/pytorch/pull/146690#discussion_r1952230796)

To integrate the `skipIfNoArm` decorator into the test file, I defined it in a way similar to `skipIfNoX86`. I then added ARM as a qengine in `torch/backends/quantized/__init__.py`. In that file, I found a note stating, "This function should correspond to the enums present in c10/core/QEngine.h," so I updated `c10/core/QEngine.h` accordingly.

Additionally, if the qconfig change was not done then,  the system defaults to the "x86" configuration, which leads to an error when we use the ARM configuration.

choudhary-devang commented:
Hi @leslie-fang-intel, replay to this comment (https://github.com/pytorch/pytorch/pull/146690#discussion_r1952402419)

To set the default qconfig as `arm` just on arm platform. if we don't set it and if backend variable is not passed in get_default_qconfig() then the function will select `x86` as default config.

Xia-Weiwen commented:
Hi @choudhary-devang Since the `qengine` and `qconfig` are deprecated and are not needed for PT2E quantization, can you please explore how to avoid changes to them while still implement the functions/features you need? For example, the implementation of `skipIfNoX86` checks `qengine` but is it possible to check if the code is running on an arm platform without using `qengine`? And similar for `qconfig`. Thanks.

abhijain1204fujitsu commented:
> Hi @choudhary-devang Since the `qengine` and `qconfig` are deprecated and are not needed for PT2E quantization, can you please explore how to avoid changes to them while still implement the functions/features you need? For example, the implementation of `skipIfNoX86` checks `qengine` but is it possible to check if the code is running on an arm platform without using `qengine`? And similar for `qconfig`. Thanks.

Thanks for sharing the feedback, will check the same.

choudhary-devang committed:
removed the QEngine and Qconfig changes

choudhary-devang commented:
> Hi @choudhary-devang Since the `qengine` and `qconfig` are deprecated and are not needed for PT2E quantization, can you please explore how to avoid changes to them while still implement the functions/features you need? For example, the implementation of `skipIfNoX86` checks `qengine` but is it possible to check if the code is running on an arm platform without using `qengine`? And similar for `qconfig`. Thanks.

Done can you please check now.

maajidkhann commented:
@pytorchbot rebase

choudhary-devang commented:
Hi @albanD, @seemethere, I am getting this (https://github.com/pytorch/pytorch/actions/runs/13434612241/job/37539552167?pr=146690) failure because of LOC limit exceed, can you please help me with this.

choudhary-devang committed:
removed trailing line

albanD commented:
Hey!
To ensure all PRs are properly reviewed and can be reliably handled by our infra, we enforce this limit on the PR size.
This PRs seems to add a lot of new public API based on a lot of new code. So I would expect it to be split up so it can be properly reviewed.

fadara01 commented on torch/ao/quantization/quantizer/arm_inductor_quantizer.py:
```diff
@@ -0,0 +1,1577 @@
+        dtype=torch.int8,
+        quant_min=-128,
+        quant_max=127,
+        qscheme=torch.per_channel_symmetric,
```
We do not currently have optimized kernels in oneDNN for `per_channel_symmetric` for AArch64 - i.e. you will just fallback to oneDNN jit/ref int8 kernels which are currently 10x slower than Arm Compute Library (ACL).

I think it might make sense to have the weight `qscheme` as `torch.per_tensor_symmetric` until we properly support `per_channel_symmetric` in oneDNN

fadara01 commented on torch/ao/quantization/quantizer/arm_inductor_quantizer.py:
```diff
@@ -0,0 +1,1577 @@
+            act_observer_or_fake_quant_ctr = HistogramObserver  # type: ignore[assignment]
+    # check for the qconfig -------------------------
+    act_quantization_spec = QuantizationSpec(
+        dtype=torch.uint8,
```
We do not currently have optimized kernels in oneDNN for `uint8_t` for AArch64. Can you use `int8_t` here.

fadara01 reviewed (changes requested):
Thank you for your PR.
I put down a few comments, but I'm generally against this getting merged in its current state.
The `arm_inductor_quantizer` this PR adds is an exact copy (word by word) of the [x86_inductor_quantizer](https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/quantizer/x86_inductor_quantizer.py) (only changing x86 occurrences to Arm). 
I do not see the point of adding 1500 new duplicate lines of code (plus tests) into PyTorch.
A better approach would be to have a general `onednn_inductor_quantizer` which x86 and Arm can extend in their `x86_inductor_quantizer` and `arm_inductor_quantizer`.

choudhary-devang committed:
added onednn_inductor_quantizer to remove redundant code in arm_inductor_quantizer and x86_inductor_quantizer

choudhary-devang commented:
> Thank you for your PR. I put down a few comments, but I'm generally against this getting merged in its current state. The `arm_inductor_quantizer` this PR adds is an exact copy (word by word) of the [x86_inductor_quantizer](https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/quantizer/x86_inductor_quantizer.py) (only changing x86 occurrences to Arm). I do not see the point of adding 1500 new duplicate lines of code (plus tests) into PyTorch. A better approach would be to have a general `onednn_inductor_quantizer` which x86 and Arm can extend in their `x86_inductor_quantizer` and `arm_inductor_quantizer`.

Hi @fadara01, thank you for your feedback. I've implemented the requested changes, could you please review them.

choudhary-devang committed:
changed function names accroding the test_public_binding conventions

maajidkhann commented:
> > Thank you for your PR. I put down a few comments, but I'm generally against this getting merged in its current state. The `arm_inductor_quantizer` this PR adds is an exact copy (word by word) of the [x86_inductor_quantizer](https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/quantizer/x86_inductor_quantizer.py) (only changing x86 occurrences to Arm). I do not see the point of adding 1500 new duplicate lines of code (plus tests) into PyTorch. A better approach would be to have a general `onednn_inductor_quantizer` which x86 and Arm can extend in their `x86_inductor_quantizer` and `arm_inductor_quantizer`.
> 
> Hi @fadara01, thank you for your feedback. I've implemented the requested changes, could you please review them.

Hi @fadara01 . Can you trigger the CI pipelines.

fadara01 commented:
@leslie-fang-intel @Xia-Weiwen - pinging you guys to re-review please. Any thoughts on the current approach this PR takes for enabling the `arm_inductor_quantizer`?

fadara01 commented:
Thanks for your refactor @choudhary-devang, there's some failing tests:
![image](https://github.com/user-attachments/assets/95c97b44-1732-4ea0-b2a5-4d38109ecde1)

choudhary-devang commented:
Hi @fadara01, replay for this comment https://github.com/pytorch/pytorch/pull/146690#discussion_r1977114706 ,
the configuration is different in case of `x86` and `arm` because we have only included annotation for `linear`, `conv` and `matmul` operation and not include any from `propagation_quantizable_ops`.

leslie-fang-intel reviewed (commented):
The CI failed.

Xia-Weiwen commented:
Hi @fadara01 I saw your refactoring of quantizers. Now `onednn_inductor_quantizer` is the parent class of `x86_inductor_quantizer` and `arm_inductor_quantizer`, right? Is it feasible that `arm_inductor_quantizer` inherits `x86_inductor_quantizer` so that we can avoid so many changes and keep `x86_inductor_quantizer` intact?

maajidkhann commented:
> Hi @fadara01 I saw your refactoring of quantizers. Now `onednn_inductor_quantizer` is the parent class of `x86_inductor_quantizer` and `arm_inductor_quantizer`, right? Is it feasible that `arm_inductor_quantizer` inherits `x86_inductor_quantizer` so that we can avoid so many changes and keep `x86_inductor_quantizer` intact?

@fadara01 I agree to [Xia-Weiwen](https://github.com/Xia-Weiwen). Even after the current refactor, the common API's are in onednn_inductor_quantizer and the rest of individual API's are in arm_inductor_quantizer and x86_inductor_quantizer.

I see there are lot of CI failures and just importing from x86_inductor_quantizer might be the best option as this would mean less changes and keep the original x86_inductor_quantizer intact.

choudhary-devang committed:
removed onednn_inductor_quantizer insted use x86_inductor_quantizer

choudhary-devang commented:
Hi @Xia-Weiwen, I have changed the structure according to your suggestions, can you please review the code once and trigger CI pipeline.

maajidkhann commented:
> Hi @Xia-Weiwen, I have changed the structure according to your suggestions, can you please review the code and trigger CI pipeline.

Can anyone trigger the CI @fadara01 @Xia-Weiwen

maajidkhann commented:
@pytorchbot rebase

maajidkhann commented:
@pytorchbot rebase -b main

Xia-Weiwen reviewed (approved):
LGTM since your change looks clean. But I haven't reviewed details of your implementation.

fadara01 commented:
@pytorchbot label "ciflow/linux-aarch64"

fadara01 commented on test/quantization/pt2e/test_arminductor_quantizer.py:
```diff
@@ -0,0 +1,1496 @@
+# Owner(s): ["oncall: quantization"]
+import copy
```
Which changes have been made here to support `per_tensor` weight quantization as the default (instead of `per_channel`)?
If no changes are needed, could you elaborate as to why.

fadara01 commented:
Do we need to add the `test_arminductor_quantizer.py` [here](https://github.com/pytorch/pytorch/blob/main/.ci/pytorch/test.sh#L1505) to make it run in CI?

cc: @robert-hardwick

choudhary-devang committed:
add the docs/source changes and adjusted the testcases for per_tensor qspec.

choudhary-devang commented:
Hi @fadara01, thanks for your efforts, I have adjusted the test_cases for the `per_tensor` change. and I think to run the test_cases in CI its need to be added it in `test_quantization.py` file i added that change.
can you please take a look  thankyou.

maajidkhann commented:
> Hi @fadara01, thanks for your efforts, I have adjusted the test_cases for the `per_tensor` change. and I think to run the test_cases in CI its need to be added it in `test_quantization.py` file i added that change. can you please take a look thankyou.

The CI is green now.
CC @jerryzh168

choudhary-devang commented:
Hi @fadara01, can you please review the changes and approve them so that it can be merged, thank you.

jerryzh168 commented on docs/source/conf.py:
```diff
@@ -2466,6 +2466,8 @@
     "X86InductorQuantizer",
     # torch.ao.quantization.quantizer.xpu_inductor_quantizer
     "XPUInductorQuantizer",
+    # torch.ao.quantization.quantizer.arm_inductor_quantizer
+    "ArmInductorQuantizer",
```
we don't want to add new things to this file I think

jerryzh168 commented:
overall looks good, how does the lowering pass work for ARM? for x86 inductor path, we use torch.compile that will finally be fusing (dq -> op -> q) patterns into onednn operators like these: https://github.com/pytorch/pytorch/blob/dee016ceb7451d2b9815ec1735afb08a6c4732e4/aten/src/ATen/native/quantized/library.cpp#L251

fadara01 commented:
> overall looks good, how does the lowering pass work for ARM? for x86 inductor path, we use torch.compile that will finally be fusing (dq -> op -> q) patterns into onednn operators like these:

@jerryzh168  lowering will also be through inductor for ArmQuantizer.
Currently this PR won't do much (i.e. models will still run fp32 GEMMs, not quantized ones) because we do not have lowering patterns in inductor for `per_tensor` weights or for (signed) `int8` inputs.
Such rules need to be implemented in [mkldnn_fusion.py](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/mkldnn_fusion.py) and [quantization.py ](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/quantization.py).

But that can be in a separate PR.

choudhary-devang committed:
removed unrequired docs changes

milpuz01 commented:
@choudhary-devang @jerryzh168 In ExecuTorch there is already Arm quantiser (https://github.com/pytorch/executorch/blob/main/backends/arm/quantizer/arm_quantizer.py) that is using TOSA as backend for quantization in order to target devices such as Ethos-U. I was wondering whether we can rename this quantiser to be `onednn_inductor_quantizer.py` as there is lot of commonality with x86 quantiser that is targeting CPU via inductor path and that path is leveraging oneDNN for efficient code?

(cc: @digantdesai @freddan80)

jerryzh168 commented:
> @choudhary-devang @jerryzh168 In ExecuTorch there is already Arm quantiser ([pytorch/executorch@`main`/backends/arm/quantizer/arm_quantizer.py](https://github.com/pytorch/executorch/blob/main/backends/arm/quantizer/arm_quantizer.py)) that is using TOSA as backend for quantization in order to target devices such as Ethos-U. I was wondering whether we can rename this quantiser to be `onednn_inductor_quantizer.py` as there is lot of commonality with x86 quantiser that is targeting CPU via inductor path and that path is leveraging oneDNN for efficient code?
> 
> (cc: @digantdesai @freddan80)

is ARM ops just (1) a different implementation of onednn ops, or (2) will they be using different hardware instructions and target different hardwares? I think we can merge into onednn if it's (1), but we should have a separate quantizer if it's (2), 

even with (2) you can compose with onednn quantizer with composable_quantizer: https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/quantizer/composable_quantizer.py and use one quantizer to quantize one part of the model and the other quantizer to quantize the other part

freddan80 commented:
> is ARM ops just (1) a different implementation of onednn ops, or (2) will they be using different hardware instructions and target different hardwares? 

@jerryzh168 Hello, good to e-meet you! 2. This quantizer is is for Arm NPUs.

I agree with @milpuz01, we should consider changing the name. Having an `ArmQuantizer` and an `ArmInductorQuantizer` will confuse ppl in the community I think. There's also XNN-pack, which has its own ´XNNpackQuantizer´, which also supports Cortex-A CPUs... Hence having `OneDNN` in the quantizer name, path etc. would make most sense to me. 

Perhaps there should be a naming convention for quantizers :) 

@digantdesai your thought on this?

choudhary-devang committed:
Revert "removed unrequired docs changes"

This reverts commit 4db960aaaf95f41c3caa6adde92927116feffe91.

freddan80 commented:
Perhaps I read the question wrong. To clarify.

[ArmQuantizer](https://github.com/pytorch/executorch/blob/main/backends/arm/quantizer/arm_quantizer.py): For NPUs
[arm_inductor_quatizer (this PR)](https://github.com/pytorch/pytorch/pull/146690/files#diff-96834df4019a286b511b4e5ed8166a1e3619ed149bdc50da2868dc819f8bc16c)  as @fadara01 point out is for Arm CPU ops behind OneDNN API's IIUC.

jerryzh168 commented:
oh I see, I thought oneDNN is just for intel cpu, in that case I think it will be better to merge into the existing X86InductorQuantizer (and should probably rename this to OnednnQuantizer), in general it can be per backend library I think, like fbgemm, onednn etc.

leslie-fang-intel commented:
> > @freddan80 nice to meet you as well, also thanks for clarifications @fadara01.
> > I thought oneDNN is just for intel cpu, in that case I think it will be better to merge into the existing X86InductorQuantizer (and should probably rename this to OnednnQuantizer), in general it can be per backend library I think, like fbgemm, onednn etc.
> 
> cc @leslie-fang-intel about the renaming suggestion (X86InductorQuantizer -> OnednnQuantizer)

Since for the backend optimization of `X86InductorQuantizer`, we will leverage both `oneDNN primitive`, `GEMM Template with X86 intrinsic` and `Inductor CPP Backend codegen`, feels like `OnednnQuantizer` may not be as intuitive as `X86InductorQuantizer`.

jerryzh168 commented:
> > > @freddan80 nice to meet you as well, also thanks for clarifications @fadara01.
> > > I thought oneDNN is just for intel cpu, in that case I think it will be better to merge into the existing X86InductorQuantizer (and should probably rename this to OnednnQuantizer), in general it can be per backend library I think, like fbgemm, onednn etc.
> > 
> > 
> > cc @leslie-fang-intel about the renaming suggestion (X86InductorQuantizer -> OnednnQuantizer)
> 
> Since for the backend optimization of `X86InductorQuantizer`, we will leverage both `oneDNN primitive`, `GEMM Template with X86 intrinsic` and `Inductor CPP Backend codegen`, feels like `OnednnQuantizer` may not be as intuitive as `X86InductorQuantizer`.

so what should name be if we add ARM CPU support on top of x86 CPU

choudhary-devang commented:
> > > > @freddan80 nice to meet you as well, also thanks for clarifications @fadara01.
> > > > I thought oneDNN is just for intel cpu, in that case I think it will be better to merge into the existing X86InductorQuantizer (and should probably rename this to OnednnQuantizer), in general it can be per backend library I think, like fbgemm, onednn etc.
> > > 
> > > 
> > > cc @leslie-fang-intel about the renaming suggestion (X86InductorQuantizer -> OnednnQuantizer)
> > 
> > 
> > Since for the backend optimization of `X86InductorQuantizer`, we will leverage both `oneDNN primitive`, `GEMM Template with X86 intrinsic` and `Inductor CPP Backend codegen`, feels like `OnednnQuantizer` may not be as intuitive as `X86InductorQuantizer`.
> 
> @leslie-fang-intel so what should name be if we add ARM CPU support on top of x86 CPU?
> 
> maybe `ServerCPU`?

Hi @jerryzh168 
As @fadara01 already mentioned above, we use few different quantization configs on ARM
compared to x86 because these configs have optimised implementations for ARM using oneDNN/ACL.
Now, I also plan to introduce further new configs and patterns for INT8 specifically for ARM
in my future PR's and these might not be applicable to x86.
 
So, I was thinking we can have a seperate ARM quantizer (arm_inductor_quantizer.py) like how 
it is this currrent PR instead of merging it into common one for ease of maintainibility and also 
for reasons mentioned above by @leslie-fang-intel

jerryzh168 commented:
@choudhary-devang OK that sounds good, we just copy pasted the pt2e quant code to torchao, could you reopen this PR in torchao instead? https://github.com/pytorch/ao/tree/main/torchao/quantization/pt2e

freddan80 commented:
> So, I was thinking we can have a seperate ARM quantizer (arm_inductor_quantizer.py) like how
it is this currrent PR instead of merging it into common one for ease of maintainibility and also
for reasons mentioned above by @leslie-fang-intel

My only concern is about naming. I think OneDNN should be there in the name somehow, or there'll be confusion. For example, XNNpack has its `XNNpackQuantizer`, which runs on Arm CPU's. To align with that naming convention, OneDNN should be in the name imo. Naming is hard - I do think `arm_inductor_quantizer` and `arm_quantizer` will be mixed up and cause confusion.

choudhary-devang commented:
> @choudhary-devang OK that sounds good, we just copy pasted the pt2e quant code to torchao, could you reopen this PR in torchao instead? https://github.com/pytorch/ao/tree/main/torchao/quantization/pt2e

Hi @jerryzh168, I have created a new pr as requested in torchao
https://github.com/pytorch/ao/pull/2139
